File: scripts/seedDummyUsers.ts
// ========================================
// Seed Script - Generate Dummy Users for Testing
// Run with: npx tsx scripts/seedDummyUsers.ts
// ========================================

import { initializeApp } from 'firebase/app';
import { getFirestore, doc, setDoc, collection, Timestamp } from 'firebase/firestore';
import { addLocationJitter, encryptLocation } from '../src/utils/securityUtils';

// Firebase config - copy from your .env or firebase.ts
const firebaseConfig = {
    apiKey: process.env.VITE_FIREBASE_API_KEY,
    authDomain: process.env.VITE_FIREBASE_AUTH_DOMAIN,
    projectId: process.env.VITE_FIREBASE_PROJECT_ID,
    storageBucket: process.env.VITE_FIREBASE_STORAGE_BUCKET,
    messagingSenderId: process.env.VITE_FIREBASE_MESSAGING_SENDER_ID,
    appId: process.env.VITE_FIREBASE_APP_ID,
};

const app = initializeApp(firebaseConfig);
const db = getFirestore(app);

// ========================================
// LOCATION PRIVACY HELPERS
// ========================================

// Obfuscate location using jitter and encryption
function obfuscateLocation(lat: number, lng: number): string {
    const jittered = addLocationJitter(lat, lng);
    return encryptLocation(jittered);
}

// ========================================
// DUMMY DATA GENERATORS
// ========================================

const FIRST_NAMES = ['Alex', 'Jordan', 'Taylor', 'Casey', 'Morgan', 'Riley', 'Quinn', 'Avery', 'Jamie', 'Skyler', 'Drew', 'Sage', 'Blake', 'Cameron', 'Dakota'];
const LAST_NAMES = ['Smith', 'Johnson', 'Williams', 'Brown', 'Jones', 'Garcia', 'Miller', 'Davis', 'Rodriguez', 'Martinez'];

const HABIT_NAMES = [
    'Morning meditation', 'Exercise', 'Read 30 mins', 'Drink 8 glasses water',
    'No social media before noon', 'Journal', 'Cold shower', 'Walk 10k steps',
    'Practice language', 'Learn something new', 'Gratitude practice', 'Stretch',
];

const TODO_TEMPLATES = [
    'Review project docs', 'Send weekly update', 'Call client', 'Update portfolio',
    'Prepare presentation', 'Research competitors', 'Fix bug #123', 'Write tests',
    'Review PR', 'Update documentation', 'Team standup notes', 'Plan sprint',
];

const GOAL_TEMPLATES = [
    { name: 'Read 24 books this year', targetValue: 24, startingValue: 0, targetDirection: 'increase', unit: 'books' },
    { name: 'Run 500 miles', targetValue: 500, startingValue: 0, targetDirection: 'increase', unit: 'miles' },
    { name: 'Save $10,000', targetValue: 10000, startingValue: 0, targetDirection: 'increase', unit: 'dollars' },
    { name: 'Lose 10kg', targetValue: 80, startingValue: 90, targetDirection: 'decrease', unit: 'kg' },
    { name: 'Reduce body fat by 5%', targetValue: 15, startingValue: 20, targetDirection: 'decrease', unit: '%' },
    { name: 'Learn Spanish', targetValue: 365, startingValue: 0, targetDirection: 'increase', unit: 'days' },
];

const RANKS = ['novice', 'apprentice', 'achiever', 'champion', 'master', 'legend'] as const;
const RANK_THRESHOLDS = { novice: 0, apprentice: 500, achiever: 2000, champion: 5000, master: 15000, legend: 50000 };

function calculateRank(xp: number): typeof RANKS[number] {
    if (xp >= RANK_THRESHOLDS.legend) return 'legend';
    if (xp >= RANK_THRESHOLDS.master) return 'master';
    if (xp >= RANK_THRESHOLDS.champion) return 'champion';
    if (xp >= RANK_THRESHOLDS.achiever) return 'achiever';
    if (xp >= RANK_THRESHOLDS.apprentice) return 'apprentice';
    return 'novice';
}

function generateId(): string {
    return Math.random().toString(36).substring(2, 15) + Math.random().toString(36).substring(2, 15);
}

function randomDate(daysAgo: number): Date {
    const date = new Date();
    date.setDate(date.getDate() - Math.floor(Math.random() * daysAgo));
    return date;
}

function randomPick<T>(arr: T[]): T {
    return arr[Math.floor(Math.random() * arr.length)];
}

// ========================================
// USER DATA GENERATORS
// ========================================

interface DummyUser {
    id: string;
    name: string;
    email: string;
    avatar?: string;
    isPro: boolean;
    isPublic: boolean;
    location?: string;
    xp: {
        totalXp: number;
        rank: string;
        xpHistory: Array<{
            id: string;
            amount: number;
            source: string;
            description: string;
            timestamp: Date;
        }>;
    };
    todos: Array<any>;
    habits: Array<any>;
    goals: Array<any>;
    sessions: Array<any>;
}

function generateDummyUser(baseLocation: { lat: number; lng: number }, index: number): DummyUser {
    const firstName = randomPick(FIRST_NAMES);
    const lastName = randomPick(LAST_NAMES);
    const name = `${firstName} ${lastName}`;
    const id = `dummy_user_${index}_${generateId()}`;

    // Generate XP (varied distribution)
    const xpDistributions = [
        () => Math.floor(Math.random() * 500), // Novice
        () => 500 + Math.floor(Math.random() * 1500), // Apprentice
        () => 2000 + Math.floor(Math.random() * 3000), // Achiever
        () => 5000 + Math.floor(Math.random() * 10000), // Champion
        () => 15000 + Math.floor(Math.random() * 35000), // Master
        () => 50000 + Math.floor(Math.random() * 20000), // Legend (rare)
    ];

    // Weighted distribution - more lower ranks
    const weights = [30, 25, 20, 15, 8, 2];
    let roll = Math.random() * 100;
    let distIndex = 0;
    for (let i = 0; i < weights.length; i++) {
        roll -= weights[i];
        if (roll <= 0) {
            distIndex = i;
            break;
        }
    }

    const totalXp = xpDistributions[distIndex]();
    const rank = calculateRank(totalXp);

    // Generate XP history
    const xpHistory = [];
    let remainingXp = totalXp;
    const sources = ['todo', 'habit', 'goal', 'milestone', 'session'];
    while (remainingXp > 0 && xpHistory.length < 50) {
        const amount = Math.min(remainingXp, Math.floor(Math.random() * 50) + 10);
        xpHistory.push({
            id: generateId(),
            amount,
            source: randomPick(sources),
            description: `Earned ${amount} XP`,
            timestamp: randomDate(30),
        });
        remainingXp -= amount;
    }

    // Generate location with offset from base (within ~30 miles)
    const locationOffset = {
        lat: (Math.random() - 0.5) * 0.8, // ~30 miles
        lng: (Math.random() - 0.5) * 0.8,
    };
    const rawLocation = {
        lat: baseLocation.lat + locationOffset.lat,
        lng: baseLocation.lng + locationOffset.lng,
    };

    // Generate todos
    const todos = [];
    const todoCount = Math.floor(Math.random() * 8) + 3;
    for (let i = 0; i < todoCount; i++) {
        todos.push({
            id: generateId(),
            text: randomPick(TODO_TEMPLATES),
            completed: Math.random() > 0.4,
            createdAt: randomDate(14),
            priority: randomPick(['high', 'medium', 'low']),
            tags: [],
        });
    }

    // Generate habits
    const habits = [];
    const habitCount = Math.floor(Math.random() * 4) + 2;
    const usedHabits = new Set<string>();
    for (let i = 0; i < habitCount; i++) {
        let habitName = randomPick(HABIT_NAMES);
        while (usedHabits.has(habitName)) {
            habitName = randomPick(HABIT_NAMES);
        }
        usedHabits.add(habitName);

        // Generate completion history
        const completedDates: string[] = [];
        for (let d = 0; d < 30; d++) {
            if (Math.random() > 0.3) { // 70% completion rate
                const date = new Date();
                date.setDate(date.getDate() - d);
                completedDates.push(date.toISOString().split('T')[0]);
            }
        }

        habits.push({
            id: generateId(),
            name: habitName,
            frequency: 'daily',
            color: `#${Math.floor(Math.random() * 16777215).toString(16).padStart(6, '0')}`,
            completedDates,
            createdAt: randomDate(60),
        });
    }

    // Generate goals
    const goals = [];
    const goalCount = Math.floor(Math.random() * 3) + 1;
    const usedGoals = new Set<string>();
    for (let i = 0; i < goalCount; i++) {
        let goalTemplate = randomPick(GOAL_TEMPLATES);
        while (usedGoals.has(goalTemplate.name)) {
            goalTemplate = randomPick(GOAL_TEMPLATES);
        }
        usedGoals.add(goalTemplate.name);

        const isDecreasing = goalTemplate.targetDirection === 'decrease';
        const range = isDecreasing
            ? goalTemplate.startingValue - goalTemplate.targetValue
            : goalTemplate.targetValue - goalTemplate.startingValue;
        const progressDelta = Math.floor(Math.random() * range * 0.8);
        const currentValue = isDecreasing
            ? goalTemplate.startingValue - progressDelta
            : goalTemplate.startingValue + progressDelta;

        goals.push({
            id: generateId(),
            name: goalTemplate.name,
            goalType: 'trackable',
            startingValue: goalTemplate.startingValue,
            targetValue: goalTemplate.targetValue,
            currentValue: currentValue,
            targetDirection: goalTemplate.targetDirection as 'increase' | 'decrease',
            unit: goalTemplate.unit,
            category: randomPick(['fitness', 'personal', 'learning', 'financial']) as any,
            tags: [],
            icon: 'target',
            color: randomPick(['#F87171', '#34D399', '#60A5FA']),
            createdAt: randomDate(30),
            deadline: new Date(Date.now() + 90 * 24 * 60 * 60 * 1000),
        });
    }

    // Generate focus sessions
    const sessions = [];
    const sessionCount = Math.floor(Math.random() * 20) + 5;
    for (let i = 0; i < sessionCount; i++) {
        const duration = randomPick([25, 25, 25, 50, 50, 75]) * 60 * 1000;
        sessions.push({
            id: generateId(),
            duration,
            completedAt: randomDate(30),
            type: 'focus',
        });
    }

    return {
        id,
        name,
        email: `${firstName.toLowerCase()}.${lastName.toLowerCase()}@example.com`,
        isPro: Math.random() > 0.7,
        isPublic: true, // All dummy users are public for testing
        location: obfuscateLocation(rawLocation.lat, rawLocation.lng),
        xp: {
            totalXp,
            rank,
            xpHistory,
        },
        todos,
        habits,
        goals,
        sessions,
    };
}

// ========================================
// SEED FUNCTION
// ========================================

async function seedDummyUsers(count: number, baseLocation: { lat: number; lng: number }) {
    console.log(`\n🌱 Seeding ${count} dummy users...\n`);

    for (let i = 0; i < count; i++) {
        const user = generateDummyUser(baseLocation, i);

        try {
            // Save user document
            await setDoc(doc(db, 'users', user.id), {
                id: user.id,
                name: user.name,
                email: user.email,
                isPro: user.isPro,
                isPublic: user.isPublic,
                location: user.location,
                xp: {
                    totalXp: user.xp.totalXp,
                    rank: user.xp.rank,
                    xpHistory: user.xp.xpHistory.slice(-10), // Only last 10 for storage efficiency
                },
                createdAt: Timestamp.now(),
            });

            // Save todos
            for (const todo of user.todos) {
                await setDoc(doc(db, 'users', user.id, 'todos', todo.id), {
                    ...todo,
                    createdAt: Timestamp.fromDate(todo.createdAt),
                });
            }

            // Save habits
            for (const habit of user.habits) {
                await setDoc(doc(db, 'users', user.id, 'habits', habit.id), {
                    ...habit,
                    createdAt: Timestamp.fromDate(habit.createdAt),
                });
            }

            // Save goals
            for (const goal of user.goals) {
                await setDoc(doc(db, 'users', user.id, 'goals', goal.id), {
                    ...goal,
                    deadline: Timestamp.fromDate(goal.deadline),
                    createdAt: Timestamp.fromDate(goal.createdAt),
                });
            }

            console.log(`✅ Created: ${user.name} (${user.xp.rank}) - ${user.xp.totalXp} XP`);
        } catch (error) {
            console.error(`❌ Failed to create ${user.name}:`, error);
        }
    }

    console.log(`\n🎉 Done! Created ${count} dummy users.\n`);
}

// ========================================
// RUN
// ========================================

// Default location: Manchester, UK (from your screenshot)
// Change this to your preferred base location
const BASE_LOCATION = {
    lat: 53.4808,
    lng: -2.2426,
};

const USER_COUNT = 15; // Adjust as needed

seedDummyUsers(USER_COUNT, BASE_LOCATION)
    .then(() => process.exit(0))
    .catch((error) => {
        console.error('Seed failed:', error);
        process.exit(1);
    });

File: scripts/seed_dummy_users.py
#!/usr/bin/env python3
"""
Seed Script - Generate Dummy Users for Testing (Python version)
Usage: python3 seed_dummy_users.py --creds path/to/service-account.json --count 15 --lat 53.48 --lng -2.24
"""

import argparse
import random
import uuid
from datetime import datetime, timedelta
import firebase_admin
from firebase_admin import credentials, firestore
from faker import Faker
import math
import base64

# ========================================
# CONFIGURATION & CONSTANTS
# ========================================

RANKS = ['novice', 'apprentice', 'achiever', 'champion', 'master', 'legend']
RANK_THRESHOLDS = {
    'novice': 0,
    'apprentice': 500,
    'achiever': 2000,
    'champion': 5000,
    'master': 15000,
    'legend': 50000
}

HABIT_NAMES = [
    'Morning meditation', 'Exercise', 'Read 30 mins', 'Drink 8 glasses water',
    'No social media before noon', 'Journal', 'Cold shower', 'Walk 10k steps',
    'Practice language', 'Learn something new', 'Gratitude practice', 'Stretch',
]

TODO_TEMPLATES = [
    'Review project docs', 'Send weekly update', 'Call client', 'Update portfolio',
    'Prepare presentation', 'Research competitors', 'Fix bug #123', 'Write tests',
    'Review PR', 'Update documentation', 'Team standup notes', 'Plan sprint',
]

GOAL_TEMPLATES = [
    {'title': 'Read 24 books this year', 'target': 24, 'unit': 'books'},
    {'title': 'Run 500 miles', 'target': 500, 'unit': 'miles'},
    {'title': 'Save $10,000', 'target': 10000, 'unit': 'dollars'},
    {'title': 'Complete 100 workouts', 'target': 100, 'unit': 'workouts'},
    {'title': 'Learn Spanish', 'target': 365, 'unit': 'days practiced'},
    {'title': 'Write 50 blog posts', 'target': 50, 'unit': 'posts'},
    {'title': 'Meditate 200 hours', 'target': 200, 'unit': 'hours'},
]

# ========================================
# LOCATION PRIVACY HELPERS
# ========================================

import hashlib
import json
import base64
import os
from cryptography.hazmat.primitives.ciphers import Cipher, algorithms, modes
from cryptography.hazmat.primitives import padding
from cryptography.hazmat.backends import default_backend

ENCRYPTION_KEY = 'arc-web-location-secret-key-12345'

def encrypt_location(data_obj):
    """
    Encrypts data to a format compatible with crypto-js.AES.encrypt(json, passphrase)
    Uses OpenSSL compatible key derivation (EVP_BytesToKey)
    """
    plaintext = json.dumps(data_obj).encode('utf-8')
    passphrase = ENCRYPTION_KEY.encode('utf-8')
    
    # Generate salt (8 bytes)
    salt = os.urandom(8)
    
    # Key derivation (EVP_BytesToKey)
    # Replicates crypto-js default behavior
    def openssl_kdf(passphrase, salt, key_len, iv_len):
        dtot = b""
        d = b""
        while len(dtot) < key_len + iv_len:
            d = hashlib.md5(d + passphrase + salt).digest()
            dtot += d
        return dtot[:key_len], dtot[key_len:key_len + iv_len]

    key, iv = openssl_kdf(passphrase, salt, 32, 16)
    
    # PKCS7 Padding
    padder = padding.PKCS7(128).padder()
    padded_data = padder.update(plaintext) + padder.finalize()
    
    # AES-256-CBC Encryption
    cipher = Cipher(algorithms.AES(key), modes.CBC(iv), backend=default_backend())
    encryptor = cipher.encryptor()
    ciphertext = encryptor.update(padded_data) + encryptor.finalize()
    
    # Format: "Salted__" + 8-byte salt + ciphertext
    result = b"Salted__" + salt + ciphertext
    return base64.b64encode(result).decode('utf-8')

def obfuscate_location(lat, lng):
    """
    Adds a ~5KM jitter to coordinates and encrypts them.
    Compatible with the web application's security layer.
    """
    jitter_km = 5
    deg_lat = jitter_km / 111.0
    deg_lng = jitter_km / (111.0 * math.cos(math.radians(lat)))

    delta_lat = (random.random() * 2 - 1) * deg_lat
    delta_lng = (random.random() * 2 - 1) * deg_lng

    jittered = {'lat': lat + delta_lat, 'lng': lng + delta_lng}
    return encrypt_location(jittered)

# ========================================
# DUMMY DATA GENERATORS
# ========================================

fake = Faker()

def calculate_rank(xp):
    if xp >= RANK_THRESHOLDS['legend']: return 'legend'
    if xp >= RANK_THRESHOLDS['master']: return 'master'
    if xp >= RANK_THRESHOLDS['champion']: return 'champion'
    if xp >= RANK_THRESHOLDS['achiever']: return 'achiever'
    if xp >= RANK_THRESHOLDS['apprentice']: return 'apprentice'
    return 'novice'

def generate_dummy_user(base_lat, base_lng, index):
    first_name = fake.first_name()
    last_name = fake.last_name()
    name = f"{first_name} {last_name}"
    user_id = f"dummy_{int(datetime.now().timestamp())}_{index}"
    
    # XP distribution - more lower ranks
    roll = random.random() * 100
    if roll < 30: xp_range = (50, 499)      # Novice
    elif roll < 55: xp_range = (500, 1999)    # Apprentice
    elif roll < 75: xp_range = (2000, 4999)   # Achiever
    elif roll < 90: xp_range = (5000, 14999)  # Champion
    elif roll < 98: xp_range = (15000, 49999) # Master
    else: xp_range = (50000, 70000)           # Legend
    
    total_xp = random.randint(*xp_range)
    rank = calculate_rank(total_xp)
    
    # Generate XP history
    xp_history = []
    sources = ['todo', 'habit', 'goal', 'milestone', 'session']
    for _ in range(10):
        xp_history.append({
            'id': str(uuid.uuid4()),
            'amount': random.randint(10, 40),
            'source': random.choice(sources),
            'description': f"Completed {random.choice(['task', 'habit', 'goal'])}",
            'timestamp': datetime.now() - timedelta(days=random.randint(0, 14))
        })
    
    # Location with offset (within ~30 miles of base)
    raw_lat = base_lat + (random.random() - 0.5) * 0.8
    raw_lng = base_lng + (random.random() - 0.5) * 0.8
    
    # Generate dummy goals
    goals = []
    for i in range(2):
        target = random.randint(50, 200)
        is_decreasing = random.random() > 0.5
        start = random.randint(target + 10, target + 100) if is_decreasing else random.randint(0, target - 10)
        current = random.randint(target, start) if is_decreasing else random.randint(start, target)
        
        goals.append({
            'id': str(uuid.uuid4()),
            'name': f"Dummy Goal {i+1}",
            'goalType': 'trackable',
            'startingValue': start,
            'targetValue': target,
            'currentValue': current,
            'targetDirection': 'decrease' if is_decreasing else 'increase',
            'unit': 'kg' if is_decreasing else 'points',
            'category': 'fitness' if is_decreasing else 'personal',
            'color': '#F87171' if is_decreasing else '#60A5FA',
            'createdAt': datetime.now() - timedelta(days=random.randint(0, 30)),
            'deadline': datetime.now() + timedelta(days=random.randint(30, 90))
        })
    
    return {
        'id': user_id,
        'name': name,
        'email': f"{first_name.lower()}.{last_name.lower()}{index}@example.com",
        'isPro': random.random() > 0.7,
        'isPublic': True,
        'location': obfuscate_location(raw_lat, raw_lng),
        'xp': {
            'totalXp': total_xp,
            'rank': rank,
            'xpHistory': xp_history
        },
        'goals': goals, # Add the generated goals here
        'createdAt': datetime.now()
    }

# ========================================
# MAIN SEED FUNCTION
# ========================================

def seed_data(creds_path, count, base_lat, base_lng):
    # Initialize Firebase Admin
    cred = credentials.Certificate(creds_path)
    firebase_admin.initialize_app(cred)
    db = firestore.client()
    
    print(f"\n🌱 Seeding {count} dummy users to project: {cred.project_id}\n")
    
    for i in range(count):
        user_data = generate_dummy_user(base_lat, base_lng, i)
        user_id = user_data.pop('id')
        
        try:
            # Save user document
            db.collection('users').document(user_id).set(user_data)
            print(f"✅ Created: {user_data['name']} ({user_data['xp']['rank']}) - {user_data['xp']['totalXp']} XP")
            
            # Additional related data (todos, habits, goals) can be added here if needed
            # For simplicity, we've bundled the key stats into the user doc for the leaderboard
            
        except Exception as e:
            print(f"❌ Failed to create {user_data['name']}: {e}")
            
    print(f"\n🎉 Done! Generated {count} users near {base_lat}, {base_lng}\n")

if __name__ == "__main__":
    parser = argparse.ArgumentParser(description='Seed Firestore with dummy user data.')
    parser.add_argument('--creds', required=True, help='Path to service-account.json')
    parser.add_argument('--count', type=int, default=15, help='Number of users to generate')
    parser.add_argument('--lat', type=float, default=53.48, help='Base latitude for location generation')
    parser.add_argument('--lng', type=float, default=-2.24, help='Base longitude for location generation')
    
    args = parser.parse_args()
    
    seed_data(args.creds, args.count, args.lat, args.lng)
